Fixing code of tau embedding (TauAnalysis/MCEmbeddingTools) for run2 needs

swayand opened the pull request:
This is a follow up PR of https://github.com/cms-sw/cmssw/pull/16271

This PR replaces https://github.com/cms-sw/cmssw/pull/16633

Commit:
Embedding filters, now added to the cc file of HepMCFilterDriver

Commit:
Changes of Embedding code in in just one commit

Commit:
Adding missing commit from PR16271

cmsbuild commented:
A new Pull Request was created by  @swayand  for CMSSW_9_0_X.

It involves the following packages:

GeneratorInterface/Core
TauAnalysis/MCEmbeddingTools


@civanch, @perrozzi, @mdhildreth, @monttj, @cmsbuild, @thuer, @govoni, @davidlange6 can you please review it and eventually sign? Thanks.
@agrohsje, @mkirsano this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f4801fc68e7a371652f1a7c860d22b3efb5a9cd5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17149/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17149/summary.html

Commit:
modified runtest, works now with CMSSW_9_0_X

cmsbuild commented:
Pull request #17095 was updated. @civanch, @perrozzi, @mdhildreth, @monttj, @cmsbuild, @thuer, @govoni, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 301c8735760501e75d47d82e2c9c4990af055189
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17154/summary.html

cmsbuild commented:
Comparison job queued.

slava77 commented:
I'm posting a copy of my signoff message from the 81X precursor PR #16271 

for https://github.com/cms-sw/cmssw/pull/17095 301c873

- reco review was added to aid with this analysis PR review due to a strong coupling with the reco software algorithm setup. 
    - Code changes look OK within the analysis scope of the MCEmbeddingTools.
- jenkins tests pass including the unit test running this workflow. There are no changes in the matrix tests.

See also some notes in https://github.com/cms-sw/cmssw/pull/16271#issuecomment-267666417 . While the detailed embedding is useful as a reference, I think that there should be a faster standard methodology relying on at least AOD to do embedding, especially in view of increasing pileup [the slow SIMembedding step could even be done with fastSim tools].

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17154/summary.html

civanch commented:
+1

civanch commented:
+1

slava77 commented:
ping
analysis and generators signatures are missing. 
It would be nice to converge before the PR is more than 3 months old (81X "parent" was submitted Oct 19).

monttj commented:
+1

perrozzi commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

davidlange6 commented on GeneratorInterface/Core/interface/BaseHadronizer.h:
```diff
@@ -111,6 +111,8 @@ namespace gen {
     std::vector<std::string> randomInitConfigDescriptions_;
     std::vector<std::string> gridpackPaths_;
+
+    char *args[5];
```
Hi @swayand- why is this change needed? It looks like this variable is used in only one spot.

davidlange6 commented on TauAnalysis/MCEmbeddingTools/plugins/DYToMuMuGenFilter.cc:
```diff
@@ -0,0 +1,125 @@
+			//std::cout << gen_particle.daughter(1)->pdgId()+gen_particle.daughter(0)->pdgId() << std::endl;
+			
+			// Check if daugther particles are muons
+		  if (abs(gen_particle.daughter(0)->pdgId()) == 13  
```
please use std::abs here.

swayand commented:
On 01/18/2017 11:25 AM, David Lange wrote:
>
> *@davidlange6* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In GeneratorInterface/Core/interface/BaseHadronizer.h 
> <https://github.com/cms-sw/cmssw/pull/17095#pullrequestreview-17189412>:
>
> > @@ -111,6 +111,8 @@ namespace gen {
>       std::vector<double> randomInitWeights_;
>       std::vector<std::string> randomInitConfigDescriptions_;
>       std::vector<std::string> gridpackPaths_;
> +
> +    char *args[5];
>
> Hi @swayand- why is this change needed? It looks like this variable is 
> used in only one spot.
>
> —
>
*@davidlange6*

I made this change because of the static code analyzer gives the 
previous implementation several critical bugs. Therefore i fixed it, 
even it has nothing to do with the embedding code itself.


> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/cms-sw/cmssw/pull/17095#pullrequestreview-17189412>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGP8_ufC8-MeWjZYoQ6ApcRipKiVan6iks5rTeiugaJpZM4LUfve>.
>

Commit:
calling abs and fabs functions from std namespace

cmsbuild commented:
Pull request #17095 was updated. @civanch, @perrozzi, @mdhildreth, @monttj, @cmsbuild, @thuer, @govoni, @davidlange6 can you please check and sign again.

swayand commented:
@davidlange6 

i made the requested fixes (added std to the abs function)

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f84364d44a45ab47bddd5392d06faeeeb501b117
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17376/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17376/summary.html

slava77 commented on TauAnalysis/MCEmbeddingTools/plugins/DYToMuMuGenFilter.cc:
```diff
@@ -74,9 +74,9 @@ bool DYToMuMuGenFilter::filter(edm::Event& iEvent, const edm::EventSetup& iSetup
-		      && fabs(gen_particle.daughter(1)->eta())<2.6
+		  if (std::abs(gen_particle.daughter(0)->pdgId()) == 13  
+		      && std::fabs(gen_particle.daughter(0)->eta())<2.6  
+		      && std::fabs(gen_particle.daughter(1)->eta())<2.6
```
better use std::abs here as well

Commit:
use std::abs now instead of fabs function

cmsbuild commented:
Pull request #17095 was updated. @civanch, @perrozzi, @mdhildreth, @monttj, @cmsbuild, @thuer, @govoni, @davidlange6 can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2999bad725d907ff95bb97dd2e1adb313a09c13f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17399/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17399/summary.html

civanch commented:
+1

perrozzi commented:
+1

swayand commented:
@monttj 
here still the analysis flag is missing. Could you check again, the changes to last time was that the abs/fabs was changed to std::abs.

davidlange6 commented:
Hi @swayand - i'm not really waiting on @monttj - rather there are some unanswered questions to you (maybe just one) and I just notice that this PR no longer merges, so needs to be rebased.

swayand commented:
@davidlange6  could you pleas point me to the unanswered question. 

Afterward i will do the rebase, or resolves the conflicts

slava77 commented:
@davidlange6 
https://github.com/cms-sw/cmssw/pull/17095#issuecomment-277953790

```
@davidlange6 could you pleas point me to the unanswered question.

Afterward i will do the rebase, or resolves the conflicts
```

swayand committed:
adding tau life time information to LHE event, since due to a bug in phythia8 this is not handeld correctly there. Also some minor code modifications are done

davidlange6 commented:
ok, you answered it but in a different part of the PR

davidlange6 commented on GeneratorInterface/Core/src/BaseHadronizer.cc:
```diff
@@ -104,7 +104,6 @@ const std::vector<std::string> BaseHadronizer::theSharedResources;
     edm::FileInPath script("GeneratorInterface/LHEInterface/data/run_generic_tarball_cvmfs.sh");
     const char *outfilename = "cmsgrid_final.lhe";
     
-    char *args[5];
```
I think a c++ way that avoids newing char*s for this is just

std::array<std::string,4>argStrs;
argStrs[0]=script.fullPath();
argStrs[1]=gridpackPath();
argStrs[2]=nevStream.str();
argStrs[3]=randomStream.str();
std::array<char*,5>args = { &argStrs[0][0], &argStrs[1][0],&argStrs[2][0],&argStrs[3][0], NULL};

then memory is managed for you

Commit:
restore version of cmssw branch

Commit:
implement suggestions from davidlange

cmsbuild commented:
Pull request #17095 was updated. @civanch, @perrozzi, @mdhildreth, @monttj, @cmsbuild, @thuer, @govoni, @davidlange6 can you please check and sign again.

Commit:
Merge branch 'CMSSW_9_0_X' into fixingforembedding_cmss90x

cmsbuild commented:
Pull request #17095 was updated. @civanch, @perrozzi, @mdhildreth, @monttj, @cmsbuild, @thuer, @govoni, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

swayand commented:
Hi, 

i have updated the PR. There are three changes between the previous version. 

1. Have updated the EmbeddingLHEProducer, because there is a bug in pyhtia8, which don't restore the tau life time. Therefore i have to add also the HepRandomEngine. The solution was already tested and works. I also contact the pythia8 team and reported about this issue.
https://github.com/swayand/cmssw/blob/4fa62e12dedb8d0f38db233ae76bce2ba810cd42/TauAnalysis/MCEmbeddingTools/plugins/EmbeddingLHEProducer.cc#L297-L300

2. I also made small code improvements, where i replaced one new allocator (where the memory wasn't freed), by just calling the function 
https://github.com/swayand/cmssw/blob/c6ee7f5d5761cabdb44025d3d1bb70600805a954/TauAnalysis/MCEmbeddingTools/plugins/EmbeddingVertexCorrector.cc#L126-L127

3. Then I implemented the suggestions from @davidlange6 and implemented std::arrays<char*> instead of char*[] 
https://github.com/swayand/cmssw/blob/f98e6e038ff93c5217f683424ebdcc616bc4aae0/GeneratorInterface/Core/src/BaseHadronizer.cc#L109-L115

cmsbuild commented:
+1
Tested at: 4fa62e12dedb8d0f38db233ae76bce2ba810cd42
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17819/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/4fa62e12dedb8d0f38db233ae76bce2ba810cd42
https://github.com/cms-sw/cmssw/commit/848b93b7cdc4bc5c0197e43e14dcd75ce171a4be
https://github.com/cms-sw/cmssw/commit/9e399c8ce8b8a54a83b9ab531e331736c69f36c6
https://github.com/cms-sw/cmssw/commit/d147abbfae62799ef6707308f48872c29a545a0d
https://github.com/cms-sw/cmssw/commit/6a075a1d60ad3fc89adfad567afe30fa8f1148bf
https://github.com/cms-sw/cmssw/commit/aeded3e5f98c3e4e73823a0e4cec843b0dbd9c1f
https://github.com/cms-sw/cmssw/commit/792a8f059a39d9db11b7990e4fce7459fdf4d62d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17819/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17819/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17095/17819/summary.html

civanch commented:
+1

davidlange6 commented:
+1

Dr15Jones commented:
@davidlange6 this fails to link in the IB
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc630/CMSSW_9_0_X_2017-02-17-2300/TauAnalysis/MCEmbeddingTools

davidlange6 commented:
right- thats what I attempted to fix for the 02-18-2300 Ib...

> On Feb 18, 2017, at 9:28 PM, Chris Jones <notifications@github.com> wrote:
> 
> @davidlange6 this fails to link in the IB
> https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc630/CMSSW_9_0_X_2017-02-17-2300/TauAnalysis/MCEmbeddingTools
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
